Userland/Libraries/LibCompress/Lzma.cpp

```diff
@@ -0,0 +1,604 @@
+    //  updates that CProb variable after decoding."
+
+    // The significance of the shift width is not explained and appears to be a magic constant.
+    constexpr size_t probability_shift_width = 5;
+
+    u32 bound = (m_range_decoder_range >> probability_bit_count) * probability;
+
+    if (m_range_decoder_code < bound) {
+        probability += ((1 << probability_bit_count) - probability) >> probability_shift_width;
+        m_range_decoder_range = bound;
+        TRY(normalize_range_decoder());
+        return 0;
+    } else {
+        probability -= probability >> probability_shift_width;
+        m_range_decoder_code -= bound;
+        m_range_decoder_range -= bound;
+        TRY(normalize_range_decoder());
+        return 1;
+    }
+}
+
+ErrorOr<u16> LzmaDecompressor::decode_symbol_using_bit_tree(size_t bit_count, Span<Probability> probability_tree)
+{
+    VERIFY(bit_count <= sizeof(u16) * 8);
+    VERIFY(probability_tree.size() >= 1ul << bit_count);
+
+    // This has been modified from the reference implementation to unlink the result and the tree index,
+    // which should allow for better readability.
+
+    u16 result = 0;
+    size_t tree_index = 1;
+
+    for (size_t i = 0; i < bit_count; i++) {
+        u16 next_bit = TRY(decode_bit_with_probability(probability_tree[tree_index]));
+        result = (result << 1) | next_bit;
+        tree_index = (tree_index << 1) | next_bit;
+    }
+
+    return result;
+}
+
+ErrorOr<u16> LzmaDecompressor::decode_symbol_using_reverse_bit_tree(size_t bit_count, Span<Probability> probability_tree)
+{
+    VERIFY(bit_count <= sizeof(u16) * 8);
+    VERIFY(probability_tree.size() >= 1ul << bit_count);
+
+    u16 result = 0;
+    size_t tree_index = 1;
+
+    for (size_t i = 0; i < bit_count; i++) {
+        u16 next_bit = TRY(decode_bit_with_probability(probability_tree[tree_index]));
+        result |= next_bit << i;
+        tree_index = (tree_index << 1) | next_bit;
+    }
+
+    return result;
+}
+
+ErrorOr<void> LzmaDecompressor::decode_literal_to_output_buffer()
+{
+    u8 previous_byte = 0;
+    if (m_total_decoded_bytes > 0) {
+        auto read_bytes = MUST(m_output_buffer.read_with_seekback({ &previous_byte, sizeof(previous_byte) }, 1));
+        VERIFY(read_bytes.size() == sizeof(previous_byte));
+    }
+
+    // "To select the table for decoding it uses the context that consists of
+    //  (lc) high bits from previous literal and (lp) low bits from value that
+    //  represents current position in outputStream."
+    u16 literal_state_bits_from_position = m_total_decoded_bytes & ((1 << m_options.literal_position_bits) - 1);
+    u16 literal_state_bits_from_output = previous_byte >> (8 - m_options.literal_context_bits);
+    u16 literal_state = literal_state_bits_from_position << m_options.literal_context_bits | literal_state_bits_from_output;
+
+    Span<Probability> selected_probability_table = m_literal_probabilities.span().slice(0x300 * literal_state, 0x300);
```

Comment:
We use this `0x300` constant in multiple places, maybe make it a constexpr constant?

Reply:
I think `probability_table_size` is fine